reactioncommerce/reaction
# Background jobs rewrite: no Meteor dependencies!

Resolves #5288
Resolves #5312
Impact: **major**  
Type: **feature**

## Changes
- New included `job-queue` plugin. It's a port of `@reactioncommerce/job-queue`, without any Meteor dependencies. Although it attempts to keep the same API, the loss of fibers means that functions that used to be magically synchronous now need to be `await`ed.
- Fully demeteorizes the `sitemap-generator` plugin. Updates its client code to use GraphQL. Updates refresh period setting to use the new `shopSettings` API.
- Removed the outdated exchange rate updater job. Setting the rates some other way should still work if you have custom plugins that rely on them.
- New APIs, more details below

### New background jobs API
The background jobs API is abstracted onto `context.backgroundJobs` so that you do not need to import from outside your own plugin to use it. (It is theoretically possible to replace the included handling with your own background job solution, if you simulate the same API exactly. This isn't tested or officially supported yet.)

To register a worker, typically done in your plugin startup function:

```js
  const workerInstance = await context.backgroundJobs.addWorker({
    type: "someUniqueJobType",
    async worker(job) {
        await doTheWork(job.data); // Whatever function you create that does the task
        job.done("Done message");
        // If anything throws, it will automatically call job.fail(errorMessage), but you
        // could also call job.fail yourself to provide better failure details.
    }
  });
```

To schedule a job for that worker, either on startup or anywhere you have context:

```js
  const job = await context.backgroundJobs.scheduleJob({
    type: "someUniqueJobType",
    data: {}, // any data your worker needs to perform the work
    // Retry is optional
    retry: {
      retries: 5,
      wait: 60000,
      backoff: "exponential"
    },
    // Priority is optional
    priority: "normal",
    // Schedule is optional if you just need to run it once.
    // Set to any text that later.js can parse.
    schedule: "every day",
    // Set cancelRepeats to true if you want to cancel all other pending jobs with the same type
    cancelRepeats: true
  });
```

To cancel all jobs that match a certain filter:

```js
await context.backgroundJobs.cancelJobs({
  type: "foo",
  data: { shopId }
});
```

To get a job by ID:

```js
const job = await context.backgroundJobs.getJob(jobId);
```

### Background job cleanup
Plugins can now request cleanup of their finished/old jobs through `registerPlugin`.

```js
app.registerPlugin({
  // ...
  backgroundJobs: {
      cleanup: [
        { type: "saveImage/local", purgeAfterDays: 7 },
        { type: "saveImage/remote", purgeAfterDays: 7 }
      ]
    }
});
```

### contextAdditions
A plugin can now add properties to context using the `contextAdditions` option when calling `registerPlugin`. They are added before any `startup` functions are run.

```js
app.registerPlugin({
  // ...
  contextAdditions: {
    something: "wicked"
  }
});
```

```js
// in startup fn or anywhere you have context
console.log(context.something); // "wicked"
```

## Breaking changes
Just one: There are no longer background jobs that keep the exchange rates up to date. Stock Reaction does not use this feature any more. If you are relying on it, everything should continue to work, but you will need to devise your own way to periodically fetch exchange rates and update the `rate` field.

Despite efforts to keep the API the same, the loss of fibers means that functions that used to be magically synchronous now need to be `await`ed. In many cases existing code would just work, but to be safe, I did not remove the non-Meteor job-collection files. Custom plugins using background jobs should switch to the new `context.backgroundJobs` API sometime before the 3.0 release, at which time the old job-queue package will be removed.

## Testing
Generally need to test anything that relies on job queue. Do all of these:
- Set up email sending and do anything that causes an email to be sent, such as placing an order or registering a new account. Verify you get the email. On Email settings page, find the email row and click retry icon. Verify you get the email again.
- Upload image files
- On shop settings page, test everything about the "Sitemaps" card. You should see logging in server console as you change the refresh period or do a manual refresh. When you view the sitemap, you should see XML.
